File: 02-Arrays-Part-II/cpp/06-count-inversions.cpp
class Solution {
public:
    int merge(vector<int>& arr, int low, int mid, int high) {
        vector<int> temp(high - low + 1);

        int left = low;
        int right = mid + 1;
        int k = 0;
        int cnt = 0;

        while (left <= mid && right <= high) {
            if (arr[left] <= arr[right]) {
                temp[k++] = arr[left++];
            } else {
                temp[k++] = arr[right++];
                cnt += (mid - left + 1);
            }
        }

        while (left <= mid) {
            temp[k++] = arr[left++];
        }

        while (right <= high) {
            temp[k++] = arr[right++];
        }

        for (int i = low; i <= high; i++) {
            arr[i] = temp[i - low];
        }

        return cnt;
    }

    int mergeSort(vector<int>& arr, int low, int high) {
        if (low >= high) {
            return 0;
        }

        int mid = low + (high - low) / 2;

        int cnt = 0;

        cnt += mergeSort(arr, low, mid);
        cnt += mergeSort(arr, mid + 1, high);
        cnt += merge(arr, low, mid, high);

        return cnt;
    }

    int numberOfInversions(vector<int>& arr) {
        return mergeSort(arr, 0, arr.size() - 1);
    }
};

File: 02-Arrays-Part-II/python/06_count_inversions.py
class Solution:
    def merge(self, arr, low, mid, high):
        temp = []
        left, right = low, mid + 1
        cnt = 0

        while left <= mid and right <= high:
            if arr[left] <= arr[right]:
                temp.append(arr[left])
                left += 1
            else:
                temp.append(arr[right])
                right += 1
                cnt += (mid - left + 1)

        while left <= mid:
            temp.append(arr[left])
            left += 1

        while right <= high:
            temp.append(arr[right])
            right += 1

        for i in range(low, high + 1):
            arr[i] = temp[i - low]

        return cnt

    def mergeSort(self, arr, low, high):
        if low >= high:
            return 0

        mid = low + (high - low) // 2

        cnt = 0
        cnt += self.mergeSort(arr, low, mid)
        cnt += self.mergeSort(arr, mid + 1, high)
        cnt += self.merge(arr, low, mid, high)

        return cnt

    def numberOfInversions(self, arr):
        return self.mergeSort(arr, 0, len(arr) - 1)
